arch/arc/core/timestamp.c

```diff
@@ -33,7 +33,7 @@ u64_t z_tsc_read(void)
 	t = (u64_t)z_tick_get();
 	count = z_arc_v2_aux_reg_read(_ARC_V2_TMR0_COUNT);
 	irq_unlock(key);
-	t *= (u64_t)sys_clock_hw_cycles_per_tick();
+	t *= (u64_t)k_ticks_to_cyc_floor32(1);
```

Comment:
Could we also clean up 32/64-bit mess? Here we are using 32-bit API to obtain 64-bit result. In z_impl_k_timer_remaining_get() we are using 64-bit API to get 32-bit result.

Why we cannot have only 64-bit version of time conversion functions? If this is really not possible, could we document where and why we are using/should use given version of the API?

Reply:
>Could we also clean up 32/64-bit mess

I agree there is a lot of 64 to 32 and back in the places updated with the new api but it could be done as separate PR to keep that one focused on api changes.

Reply:
That spot actually looks like a clear pre-existing bug to me.  Up-casting the result of a 32 bit conversion instead of using the 64 bit one is clearly not the intent of the code, it's just that it didn't have a full precision version before and it does now.  I'll clean that up in a separate commit and push it to the end of this PR, sure (at the risk, obviously, of doing "too much")

I'm neutral on the "only 64 bit" argument.  Obviously the reason it's done the way it is now is for performance, though changing things wouldn't be too awful I suspect.  Remember that if (1) the result is a 32 bit quantity and (2) the two frequencies are integer multiples of each other, that the result can be had with perfect precision using a single 32 bit operation.   This covers almost all of our platforms (not nRF, though) and almost all of the existing usage in the code base.  I think there's value in keeping it personally, but not so much that I care to fight over it.  @carlescufi @nashif you guys make the call.

Reply:
I am partial to cleaning up and settling on the '64-bit only' solution, but I also agree with @nordic-krch that this should be a separate PR to this particular one, in order to actually gauge the support for it cleanly.